File: src/toy-machines.ts
import { type TuringMachineSpec } from "./types";

export const write1sForeverSpec: TuringMachineSpec<"init", "_" | "1"> = {
  allStates: ["init"] as const,
  allSymbols: ["_", "1"] as const,
  initial: "init",
  blank: "_",
  acceptingStates: new Set(["init"]),
  rules: new Map([
    [
      "init",
      new Map([
        ["_", ["init", "1", "R"]],
        ["1", ["init", "1", "R"]],
      ]),
    ],
  ]),
};

export const acceptImmediatelySpec: TuringMachineSpec<"init", "_"> = {
  allStates: ["init"] as const,
  allSymbols: ["_"] as const,
  initial: "init",
  blank: "_",
  acceptingStates: new Set(["init"]),
  rules: new Map([]),
};

export const rejectImmediatelySpec: TuringMachineSpec<"init", "_"> = {
  allStates: ["init"] as const,
  allSymbols: ["_"] as const,
  initial: "init",
  acceptingStates: new Set([]),
  blank: "_",
  rules: new Map(),
};

export const flipBitsSpec: TuringMachineSpec<"init", "_" | "0" | "1"> = {
  allStates: ["init"] as const,
  allSymbols: ["_", "0", "1"] as const,
  initial: "init",
  blank: "_",
  acceptingStates: new Set(["init"]),
  rules: new Map([
    [
      "init",
      new Map([
        ["0", ["init", "1", "R"]],
        ["1", ["init", "0", "R"]],
      ]),
    ],
  ]),
};

type Letter =
  | "a"
  | "b"
  | "c"
  | "d"
  | "e"
  | "f"
  | "g"
  | "h"
  | "i"
  | "j"
  | "k"
  | "l"
  | "m"
  | "n"
  | "o"
  | "p"
  | "q"
  | "r"
  | "s"
  | "t"
  | "u"
  | "v"
  | "w"
  | "x"
  | "y"
  | "z";

type PalindromeSymbol = "_" | Letter;
type PalindromeState =
  | "start"
  | "accept"
  | "seekL"
  | `seekR_${Letter}`
  | `check_${Letter}`;

export const checkPalindromeSpec = ((): TuringMachineSpec<
  PalindromeState,
  PalindromeSymbol
> => {
  // The classic approach: repeatedly match & erase the outermost pair of characters.
  //
  // States:
  //   start     – read (and erase) the leftmost symbol
  //   seekR_x   – scan right to find the end, carrying letter x
  //   check_x   – at the rightmost symbol, verify it matches x
  //   seekL     – scan left back to the start
  //   accept    – halting accept state

  type Letter =
    | "a"
    | "b"
    | "c"
    | "d"
    | "e"
    | "f"
    | "g"
    | "h"
    | "i"
    | "j"
    | "k"
    | "l"
    | "m"
    | "n"
    | "o"
    | "p"
    | "q"
    | "r"
    | "s"
    | "t"
    | "u"
    | "v"
    | "w"
    | "x"
    | "y"
    | "z";

  type PalindromeSymbol = "_" | Letter;
  type PalindromeState =
    | "start"
    | "accept"
    | "seekL"
    | `seekR_${Letter}`
    | `check_${Letter}`;
  type Dir = "L" | "R";

  const letters: readonly Letter[] = [
    ..."abcdefghijklmnopqrstuvwxyz",
  ] as Letter[];

  const allSymbols: readonly PalindromeSymbol[] = ["_", ...letters];

  const allStates: readonly PalindromeState[] = [
    "start",
    "accept",
    "seekL",
    ...letters.map((l): `seekR_${Letter}` => `seekR_${l}`),
    ...letters.map((l): `check_${Letter}` => `check_${l}`),
  ];

  function buildRules(): ReadonlyMap<
    PalindromeState,
    ReadonlyMap<PalindromeSymbol, [PalindromeState, PalindromeSymbol, Dir]>
  > {
    const rules = new Map<
      PalindromeState,
      Map<PalindromeSymbol, [PalindromeState, PalindromeSymbol, Dir]>
    >();

    // ── start: read leftmost symbol, erase it, seek right ──
    const startRules = new Map<
      PalindromeSymbol,
      [PalindromeState, PalindromeSymbol, Dir]
    >();
    startRules.set("_", ["accept", "_", "R"]); // empty tape → palindrome
    for (const l of letters) {
      startRules.set(l, [`seekR_${l}`, "_", "R"]); // erase & remember
    }
    rules.set("start", startRules);

    // ── seekR_x: scan right past all letters until blank ──
    for (const l of letters) {
      const m = new Map<
        PalindromeSymbol,
        [PalindromeState, PalindromeSymbol, Dir]
      >();
      m.set("_", [`check_${l}`, "_", "L"]); // hit the end, step back
      for (const l2 of letters) {
        m.set(l2, [`seekR_${l}`, l2, "R"]); // keep scanning
      }
      rules.set(`seekR_${l}`, m);
    }

    // ── check_x: verify rightmost symbol matches x ──
    for (const l of letters) {
      const m = new Map<
        PalindromeSymbol,
        [PalindromeState, PalindromeSymbol, Dir]
      >();
      m.set("_", ["accept", "_", "R"]); // single char was left → ok
      m.set(l, ["seekL", "_", "L"]); // match → erase & go back
      // any other letter → no rule → halt & reject
      rules.set(`check_${l}`, m);
    }

    // ── seekL: scan left back to the erased region ──
    const seekLRules = new Map<
      PalindromeSymbol,
      [PalindromeState, PalindromeSymbol, Dir]
    >();
    seekLRules.set("_", ["start", "_", "R"]); // reached left edge, restart
    for (const l of letters) {
      seekLRules.set(l, ["seekL", l, "L"]); // keep scanning
    }
    rules.set("seekL", seekLRules);

    return rules;
  }

  return {
    allStates,
    allSymbols,
    initial: "start",
    blank: "_",
    acceptingStates: new Set<PalindromeState>(["accept"]),
    rules: buildRules(),
  };
})();

type DoubleXSymbol = "_" | "$" | "X" | "Y" | "Z";
type DoubleXState =
  | "start"
  | "findX"
  | "goRight"
  | "goBack"
  | "cleanL"
  | "cleanR"
  | "done";

export const doubleXSpec: TuringMachineSpec<DoubleXState, DoubleXSymbol> =
  (() => {
    // Doubles a string of X's preceded by $.
    // Algorithm:
    //   1. For each X, mark it as Y and append a Z at the right end.
    //   2. When no unmarked X's remain, convert all Y's and Z's back to X's.
    //
    // States:
    //   start   – skip past $
    //   findX   – find leftmost unmarked X
    //   goRight – scan right to append a Z
    //   goBack  – return left to find next X
    //   cleanL  – convert Y→X going left
    //   cleanR  – convert Z→X going right
    //   done    – halting accept state

    type S = DoubleXSymbol;
    type Q = DoubleXState;
    type Dir = "L" | "R";
    type Rule = [Q, S, Dir];

    const rules = new Map<Q, Map<S, Rule>>();

    rules.set("start", new Map<S, Rule>([["$", ["findX", "$", "R"]]]));

    rules.set(
      "findX",
      new Map<S, Rule>([
        ["X", ["goRight", "Y", "R"]], // mark & go append
        ["Y", ["findX", "Y", "R"]], // skip marked
        ["Z", ["cleanL", "Z", "L"]], // all X's marked, clean up
        ["_", ["done", "_", "L"]], // no X's at all (0 case)
      ]),
    );

    rules.set(
      "goRight",
      new Map<S, Rule>([
        ["X", ["goRight", "X", "R"]], // skip remaining X's
        ["Z", ["goRight", "Z", "R"]], // skip already-appended Z's
        ["_", ["goBack", "Z", "L"]], // append Z, head back
      ]),
    );

    rules.set(
      "goBack",
      new Map<S, Rule>([
        ["X", ["goBack", "X", "L"]],
        ["Y", ["goBack", "Y", "L"]],
        ["Z", ["goBack", "Z", "L"]],
        ["$", ["findX", "$", "R"]], // back at start, find next X
      ]),
    );

    rules.set(
      "cleanL",
      new Map<S, Rule>([
        ["Y", ["cleanL", "X", "L"]], // convert Y→X
        ["$", ["cleanR", "$", "R"]], // reached $, now go right
      ]),
    );

    rules.set(
      "cleanR",
      new Map<S, Rule>([
        ["X", ["cleanR", "X", "R"]], // skip converted X's
        ["Z", ["cleanR", "X", "R"]], // convert Z→X
        ["_", ["done", "_", "L"]], // finished
      ]),
    );

    return {
      allStates: [
        "start",
        "findX",
        "goRight",
        "goBack",
        "cleanL",
        "cleanR",
        "done",
      ],
      allSymbols: ["_", "$", "X", "Y", "Z"],
      initial: "start",
      blank: "_",
      acceptingStates: new Set<Q>(["done"]),
      rules,
    };
  })();

File: src/toy_machines.rs
// ════════════════════════════════════════════════════════════════════
// Toy machines for testing the UTM
// ════════════════════════════════════════════════════════════════════

use std::{
    collections::{BTreeMap, BTreeSet},
    sync::LazyLock,
};

use crate::tm::{Dir, SimpleTuringMachineSpec};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum AccImmState {
    Init,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum AccImmSymbol {
    Blank,
}
pub static ACCEPT_IMMEDIATELY_SPEC: LazyLock<SimpleTuringMachineSpec<AccImmState, AccImmSymbol>> =
    LazyLock::new(|| SimpleTuringMachineSpec {
        initial: AccImmState::Init,
        blank: AccImmSymbol::Blank,
        accepting: BTreeSet::from([AccImmState::Init]),
        transitions: BTreeMap::new(),
        all_states: vec![AccImmState::Init],
        all_symbols: vec![AccImmSymbol::Blank],
    });

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum RejImmState {
    Init,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum RejImmSymbol {
    Blank,
}
pub static REJECT_IMMEDIATELY_SPEC: LazyLock<SimpleTuringMachineSpec<RejImmState, RejImmSymbol>> =
    LazyLock::new(|| SimpleTuringMachineSpec {
        initial: RejImmState::Init,
        blank: RejImmSymbol::Blank,
        accepting: BTreeSet::new(),
        transitions: BTreeMap::new(),
        all_states: vec![RejImmState::Init],
        all_symbols: vec![RejImmSymbol::Blank],
    });

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Letter {
    A,
    B,
    C,
    D,
    E,
    F,
    G,
    H,
    I,
    J,
    K,
    L,
    M,
    N,
    O,
    P,
    Q,
    R,
    S,
    T,
    U,
    V,
    W,
    X,
    Y,
    Z,
}
fn all_letters() -> [Letter; 26] {
    [
        Letter::A,
        Letter::B,
        Letter::C,
        Letter::D,
        Letter::E,
        Letter::F,
        Letter::G,
        Letter::H,
        Letter::I,
        Letter::J,
        Letter::K,
        Letter::L,
        Letter::M,
        Letter::N,
        Letter::O,
        Letter::P,
        Letter::Q,
        Letter::R,
        Letter::S,
        Letter::T,
        Letter::U,
        Letter::V,
        Letter::W,
        Letter::X,
        Letter::Y,
        Letter::Z,
    ]
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum CheckPalindromeState {
    Start,
    Accept,
    SeekL,
    SeekR(Letter),
    Check(Letter),
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum CheckPalindromeSymbol {
    Blank,
    Letter(Letter),
}

pub static CHECK_PALINDROME_SPEC: LazyLock<
    SimpleTuringMachineSpec<CheckPalindromeState, CheckPalindromeSymbol>,
> = LazyLock::new(|| {
    use CheckPalindromeState::*;
    use CheckPalindromeSymbol::*;

    let mut transitions: BTreeMap<
        (CheckPalindromeState, CheckPalindromeSymbol),
        (CheckPalindromeState, CheckPalindromeSymbol, Dir),
    > = BTreeMap::new();

    // start rules
    transitions.insert((Start, Blank), (Accept, Blank, Dir::Right));
    for letter in all_letters() {
        transitions.insert((Start, Letter(letter)), (SeekR(letter), Blank, Dir::Right));
    }

    for letter in all_letters() {
        transitions.insert((SeekR(letter), Blank), (Check(letter), Blank, Dir::Left));
        for l2 in all_letters() {
            transitions.insert(
                (SeekR(letter), Letter(l2)),
                (SeekR(letter), Letter(l2), Dir::Right),
            );
        }
    }

    // check_x rules
    for letter in all_letters() {
        transitions.insert((Check(letter), Blank), (Accept, Blank, Dir::Right));
        transitions.insert((Check(letter), Letter(letter)), (SeekL, Blank, Dir::Left));
    }

    // seekL rules
    transitions.insert((SeekL, Blank), (Start, Blank, Dir::Right));
    for letter in all_letters() {
        transitions.insert((SeekL, Letter(letter)), (SeekL, Letter(letter), Dir::Left));
    }

    SimpleTuringMachineSpec {
        initial: CheckPalindromeState::Start,
        accepting: BTreeSet::from([CheckPalindromeState::Accept]),
        blank: CheckPalindromeSymbol::Blank,
        transitions: BTreeMap::from(transitions),
        all_states: {
            let mut v = vec![
                CheckPalindromeState::Start,
                CheckPalindromeState::Accept,
                CheckPalindromeState::SeekL,
            ];
            for letter in all_letters() {
                v.push(CheckPalindromeState::SeekR(letter));
                v.push(CheckPalindromeState::Check(letter));
            }
            v
        },
        all_symbols: {
            let mut v = vec![CheckPalindromeSymbol::Blank];
            for letter in all_letters() {
                v.push(CheckPalindromeSymbol::Letter(letter));
            }
            v
        },
    }
});

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum DoubleXState {
    Start,
    FindX,
    GoRight,
    GoBack,
    CleanL,
    CleanR,
    Done,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum DoubleXSymbol {
    Blank,
    Dollar,
    X,
    Y,
    Z,
}
pub static DOUBLE_X_SPEC: LazyLock<SimpleTuringMachineSpec<DoubleXState, DoubleXSymbol>> =
    LazyLock::new(|| {
        use DoubleXState::*;
        use DoubleXSymbol::*;
        SimpleTuringMachineSpec {
            initial: DoubleXState::Start,
            accepting: BTreeSet::from([DoubleXState::Done]),
            blank: DoubleXSymbol::Blank,
            transitions: BTreeMap::from([
                ((Start, Dollar), (FindX, Dollar, Dir::Right)),
                ((FindX, X), (GoRight, Y, Dir::Right)),
                ((FindX, Y), (FindX, Y, Dir::Right)),
                ((FindX, Z), (CleanL, Z, Dir::Left)),
                ((FindX, Blank), (Done, Blank, Dir::Left)),
                ((GoRight, X), (GoRight, X, Dir::Right)),
                ((GoRight, Z), (GoRight, Z, Dir::Right)),
                ((GoRight, Blank), (GoBack, Z, Dir::Left)),
                ((GoBack, X), (GoBack, X, Dir::Left)),
                ((GoBack, Y), (GoBack, Y, Dir::Left)),
                ((GoBack, Z), (GoBack, Z, Dir::Left)),
                ((GoBack, Dollar), (FindX, Dollar, Dir::Right)),
                ((CleanL, Y), (CleanL, X, Dir::Left)),
                ((CleanL, Dollar), (CleanR, Dollar, Dir::Right)),
                ((CleanR, X), (CleanR, X, Dir::Right)),
                ((CleanR, Z), (CleanR, X, Dir::Right)),
                ((CleanR, Blank), (Done, Blank, Dir::Left)),
            ]),
            all_states: vec![Start, FindX, GoRight, GoBack, CleanL, CleanR, Done],
            all_symbols: vec![Blank, Dollar, X, Y, Z],
        }
    });

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FlipBitsState {
    Flip,
    Done,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum FlipBitsSymbol {
    Blank,
    Zero,
    One,
}
pub static FLIP_BITS_SPEC: LazyLock<SimpleTuringMachineSpec<FlipBitsState, FlipBitsSymbol>> =
    LazyLock::new(|| {
        use FlipBitsState::*;
        use FlipBitsSymbol::*;
        SimpleTuringMachineSpec {
            initial: Flip,
            accepting: BTreeSet::from([Done]),
            blank: Blank,
            transitions: BTreeMap::from([
                ((Flip, Zero), (Flip, One, Dir::Right)),
                ((Flip, One), (Flip, Zero, Dir::Right)),
                ((Flip, Blank), (Done, Blank, Dir::Left)),
            ]),
            all_states: vec![Flip, Done],
            all_symbols: vec![Blank, Zero, One],
        }
    });

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tm::{run_tm, HaltReason, RunningTuringMachine};

    #[test]
    fn accept_immediately() {
        use AccImmSymbol::*;
        let mut tm = RunningTuringMachine::new(&*ACCEPT_IMMEDIATELY_SPEC);
        tm.tape = vec![Blank];
        let result = run_tm(&mut tm, 1000, None);
        assert!(matches!(result, Ok(HaltReason::Accepted { .. })));
    }

    #[test]
    fn reject_immediately() {
        use RejImmSymbol::*;
        let mut tm = RunningTuringMachine::new(&*REJECT_IMMEDIATELY_SPEC);
        tm.tape = vec![Blank];
        let result = run_tm(&mut tm, 1000, None);
        assert!(matches!(result, Ok(HaltReason::Rejected { .. })));
    }

    #[test]
    fn double_x() {
        use DoubleXSymbol::*;
        let mut tm = RunningTuringMachine::new(&*DOUBLE_X_SPEC);
        tm.tape = vec![Dollar, X, X, X];
        let result = run_tm(&mut tm, 1000, None);
        assert!(matches!(result, Ok(HaltReason::Accepted { .. })));
        while tm.tape.last() == Some(&Blank) {
            tm.tape.pop();
        }
        assert_eq!(tm.tape, vec![Dollar, X, X, X, X, X, X]);
    }

    #[test]
    fn check_palindrome_accepts_odd_length_palindrome() {
        use super::Letter::*;
        use CheckPalindromeSymbol::*;
        let mut tm = RunningTuringMachine::new(&*CHECK_PALINDROME_SPEC);
        tm.tape = [A, B, C, B, A].map(Letter).to_vec();
        let result = run_tm(&mut tm, 1000, None);
        assert!(matches!(result, Ok(HaltReason::Accepted { .. })));
    }

    #[test]
    fn check_palindrome_accepts_even_length_palindrome() {
        use super::Letter::*;
        use CheckPalindromeSymbol::*;
        let mut tm = RunningTuringMachine::new(&*CHECK_PALINDROME_SPEC);
        tm.tape = [A, B, B, A].map(Letter).to_vec();
        let result = run_tm(&mut tm, 1000, None);
        assert!(matches!(result, Ok(HaltReason::Accepted { .. })));
    }

    #[test]
    fn check_palindrome_rejects_non_palindrome() {
        use super::Letter::*;
        use CheckPalindromeSymbol::*;
        let mut tm = RunningTuringMachine::new(&*CHECK_PALINDROME_SPEC);
        tm.tape = [A, B, C, A].map(Letter).to_vec();
        let result = run_tm(&mut tm, 1000, None);
        assert!(matches!(result, Ok(HaltReason::Rejected { .. })));
    }
}
